addons/hi-res-thumbnails/userscript.js

```diff
@@ -0,0 +1,66 @@
+export default async function ({ addon, console }) {
+  main: while (true) {
+    const image = await addon.tab.waitForElement("img", {
+      markAsSeen: true,
+    });
+
+    const lazy = image.classList.contains("lazy");
+    const src = lazy ? image.dataset.original : image.src;
+
+    let width, height, newSrc;
+
+    const cdn2 = src.match(/(.*\/get_image\/.*?\/[0-9]+?_)([0-9]+?)x([0-9]+?)(\.[a-z]+)/);
+
+    if (cdn2) {
+      width = cdn2[2];
+      height = cdn2[3];
+    } else {
+      continue main;
+    }
+
+    if (addon.settings.get("fixLowRes"))
+      fixLowRes: {
+        // Note this returns a value in CSS pixel units, i.e. it doesn't take into
+        // account browser scaling or high-DPI screens. That's good, because we
+        // want to leave the actual resolution-increase factor completely up to
+        // the addon user/settings!
+        const rect = image.getBoundingClientRect();
+
+        // Use greater of both scales to choose the dimension which needs the most
+        // scaling and to maintain the ratio between the rendered and loaded
+        // dimensions.
+        const scale = Math.max(rect.width / width, rect.height / height);
+
+        if (scale > 1) {
+          width *= scale;
+          height *= scale;
+        }
+      }
+
+    [width, height] = scaleDimensions(width, height);
+
+    if (cdn2) {
+      newSrc = cdn2[1] + width + "x" + height + cdn2[4];
+    }
+
+    if (lazy) {
+      image.dataset.original = newSrc;
+    }
+
+    if (image.src) {
+      image.src = newSrc;
+    }
+  }
+
+  function scaleDimensions(width, height) {
+    const factor = 0.01 * addon.settings.get("multiplier");
```

Comment:
I decided to make this a setting instead of trying to detect the browser zoom level (or "real" image render size) in some way. Partly because 1) [it's really difficult (possibly impossible) to actually detect that](https://css-tricks.com/can-javascript-detect-the-browsers-zoom-level/), but more importantly because 2) that either wouldn't work with dynamic zooming (i.e. as user zooms further in, image becomes blurry again), or it *would* detect zooming and would cause a TON of slightly-different-resolution image requests as the user adjusts their zoom level.

Leaving it up to the user solves both those problems while also giving them specific control over how big their thumbnails are going to end up, allowing for a maximum "nice" zoom and letting the user decide what they want that to be.

Reply:
I think `devicePixelRatio` and not the zoom level is what you would need, but that still doesn't solve problem 2.

Reply:
Yeah, `devicePixelRatio` is *theoretically* the solution, but it [isn't consistently supported with regards to zooming](https://css-tricks.com/can-javascript-detect-the-browsers-zoom-level/#there-is-pixel-density).

Another point against dynamically detecting zoom is that users might deliberately zoom further in for the sake of reading *text* more easily (or inspecting smaller layout details)—if a user doesn't care about the image resolution here, currently they can just leave the resolution setting at the max value they deem relevant, instead of loading a bunch of data unnecessarily.

Problem 2 could also cause a lot of lag (replacing image sources frequently), and either we update *all* images (= even more lag and data usage) or we lazily target only images on- or near-screen (= less lag but pop-in as you scroll, and even more code complexity on top of dynamic zoom detection and image updating in the first place).